File: madameautomata-bot-python-autoreg/bot/commands/starchart.py

import discord
from discord import app_commands, Interaction
from ..theme import gothic_embed
from ..util.perm import is_domme, is_staff, is_sub
from ..db import Session, User, Collaring, Cage, JailSession, Star, Task, Worship, Config, CollarType, StarStatus
from ..util.time import parse_duration, random_case
from ..util.ids import cuid
from datetime import datetime, timedelta
from sqlalchemy import select, func, update, and_

def setup(bot):
    @bot.tree.command(name="starchart", description="Star leaderboards (all-time)")
    async def starchart(i: Interaction):
        async with Session() as s:
            rows = (await s.execute(
                select(Star.sub_id, func.coalesce(func.sum(Star.amount),0))
                .where(Star.status==StarStatus.APPROVED)
                .group_by(Star.sub_id)
                .order_by(func.sum(Star.amount).desc())
                .limit(10)
            )).all()
        if not rows:
            return await i.response.send_message(embed=gothic_embed("All-Time Starchart", "—"))
        lines = [f"**{idx+1}.** <@{sid}> — {amt} ⭐" for idx,(sid,amt) in enumerate(rows)]
        await i.response.send_message(embed=gothic_embed("All-Time Starchart", "\n".join(lines)))

File: madameautomata-js-bot/src/commands/starchart.js

import { SlashCommandBuilder, PermissionFlagsBits, ChannelType, ActionRowBuilder, ButtonBuilder, ButtonStyle, StringSelectMenuBuilder, EmbedBuilder } from "discord.js";
import { prisma } from "../lib/prisma.js";
import { Theme, gothicEmbed } from "../theme/theme.js";
import { parseDuration, randomCase } from "../util/duration.js";
import { cuid } from "../util/ids.js";
import { isDomme, isSub, isStaff, getConfig } from "../util/perm.js";

const cmd = 
{
  data: new SlashCommandBuilder().setName("starchart").setDescription("All-time starchart"),
  async run(i) {
    const rows = await prisma.star.groupBy({
      by: ["subId"],
      _sum: { amount: true },
      where: { status: "APPROVED" },
      orderBy: { _sum: { amount: "desc" } },
      take: 10
    });
    if (!rows.length) return i.reply({ embeds: [gothicEmbed("All-Time Starchart", "—")] });
    const lines = rows.map((r,idx)=>`**${idx+1}.** <@${r.subId}> — ${r._sum.amount} ⭐`).join("\n");
    return i.reply({ embeds: [gothicEmbed("All-Time Starchart", lines)] });
  }
}


export default cmd;
